doc/dev/kdb-contracts.md

```diff
@@ -0,0 +1,43 @@
+# KDB Contracts
+
+The `kdbOpen()` function accepts a `KeySet * contract` parameter.
+This parameter allows you to extend and configure the KDB within your application without permanently changing files on disk.
+
+## Contract Structure
+
+The contract consists of Keys below `system:/elektra/contract/<type>`, where `<type>` is one of a set of predefined contract types.
+Currently, the types `globalkeyset` and `mountglobal` are supported.
+
+### Global KeySet Contracts
+
+To insert data into the global KeySet during `kdbOpen`, you can add keys below `system:/elektra/contract/globalkeyset`.
+All these keys will be renamed and copied into the global KeySet that is passed to all plugins.
+
+Specifically, the subset below `system:/elektra/contract/globalkeyset` is moved to `system:/elektra` and then inserted into the global KeySet.
+
+### Mounting Global Plugins
+
+You can also mount a global plugin with a contract.
+To do this, add a key `system:/elektra/contract/mountglobal/<plugin>` where `<plugin>` is the name of the plugin you want to mount.
+The keys below `system:/elektra/contract/mountglobal/<plugin>` will be moved to `user:/` and used as the config for `<plugin>`.
+
+## Pre-defined Contracts
+
+There are a few pre-defined contracts that can be accessed via helper functions.
+
+### GOpts
+
+To mount and configure the `gopts` plugin you can use `elektraGOptsContract`.
+There is also `elektraGOptsContractFromStrings`, but it's use is discouraged unless you really need it.
```

Comment:
We would have to move the whole implementation into C++ and since that is header-only and inline, I don't think that would be a good idea. The only other option I can think of right now would be to remove it from `kdb.h` and use an `extern` declaration in `C++` to call it.

However, there could be use cases were calling `elektraGOptsContractFromStrings` from C makes sense. `elektraGOptsContract` requires that the pointers are valid for a long time, this doesn't matter if you use the `argv` from `main` and the global `environ`. But if you write some kind of wrapper (e.g. if something like the FUSE binding, but directly in C), then you might want to use `elektraGOptsContractFromStrings`. It copies the data and you can clean up your pointers right away to avoid memory leaks.

Reply:
If we have elektraGOptsContract.* externalized it is not such a big deal anymore.